### ISSUE TITLE
extra dimension for jacobian of model, or jacfwd on model.apply

### ISSUE THREAD

**Issue by dwidemann** [deepmind/dm-haiku#35]
Hi:

I'm computing the Jacobian of my model on every step. The input is a z-dimensional vector and model's output is (bz,ch,h,w), it's a decoder. I expected the output of the Jacobian to be (bz,ch,h,w,z) and it is when i compute the jacobian of a row in my batch. However, when I stick in the whole batch, the dimensions become: (bz,ch,h,w,bz,z). Why the extra bz? 

I checked the Jacobian output values for the i^th row, and (i,ch,h,w,j,z) is zero for all j != i. And it has the correct values when j = i. I can still use this; however, I have to take the extra step of removing the zero outputs.

Here's my code for the Jacobian:

**decoder_jac = hk.transform(lambda z: jacfwd(Decoder())(z))**

**Comment by tomhennigan** [comment]
Hi @dwidemann, it's not immediately obvious to me what is wrong here, but one observation is that JAX function transformations (e.g. `jacfwd`) are only "safe" to use on pure functions (e.g. the `init` or `apply` function you get back from `hk.transform`). Can you try moving the jacfwd around that:

```python
decoder = hk.transform(lambda z: Decoder()(z))
decoder_jac = jacfwd(decoder.apply)
```

If not it would be very helpful if you could make a [colab](https://colab.research.google.com/) with a minimal reproduction of your issue so we can figure it out together.

**Comment by dwidemann** [comment]
Hi @tomhennigan :

Thank you for getting back to me. I tried your suggestion and it's giving me the same output. Here's a [colab](https://colab.research.google.com/gist/dwidemann/c8ff6caaa9dea2c976e24b30deca4503/haiku_jacobian_example.ipynb) of the issue. 

Thanks,

-Dave

**Comment by trevorcai** [comment]
Sorry we dropped the ball on this issue.

I think this is intentional from JAX's perspective:
for a function that takes input of shape `[B, Z]` and returns shape `[B, H, W]`, the full Jacobian should indeed be of shape `[B*H*W, B*Z]`, which is helpfully reshaped to `[B, H, W, B, Z]`.

This is because JAX doesn't know that each batch's value doesn't impact the others a priori.

Some Haiku modules, like`hk.Linear` in your example, can operate on unbatched inputs.
If you change your module to accept unbatched inputs (e.g. by changing the reshape to be `jnp.reshape(logits, self._output_shape)`), then this will work for you (and probably be more efficient):

```
jax.vmap(jax.jacfwd(decoder.apply, argnums=1), in_axes=(None, 0))(params, z)
```

This takes advantage of JAX's automatic batching via `jax.vmap` to express a batched computation of Jacobians per example, rather than computing the Jacobian of a batched computation.

You allude to channels, height, and width, suggesting your real decoder might plausibly be using transposed convolutions, which currently doesn't have the unbatched case implemented.

This can be ameliorated by just doing:
```
def f_single(params, z):
  return decoder.apply(params, z[None, ...])[0]
```
... and using the vmap trick as before.

**Comment by dwidemann** [comment]
Thank you. That works. I implemented your suggestions [here](https://colab.research.google.com/gist/dwidemann/c8ff6caaa9dea2c976e24b30deca4503/haiku_jacobian_example.ipynb#scrollTo=QhhVlgzjtk5R). It's in the **vmap solution** section. Thanks!
